util/ranger/ranger.go

```diff
@@ -410,23 +412,38 @@ func hasPrefix(lengths []int) bool {
 	return false
 }
 
-func fixPrefixColRange(ranges []*Range, lengths []int) {
+func fixPrefixColRange(ranges []*Range, lengths []int, tp []*types.FieldType) {
 	for _, ran := range ranges {
 		for i := 0; i < len(ran.LowVal); i++ {
-			fixRangeDatum(&ran.LowVal[i], lengths[i])
+			fixRangeDatum(&ran.LowVal[i], lengths[i], tp[i])
 		}
 		ran.LowExclude = false
 		for i := 0; i < len(ran.HighVal); i++ {
-			fixRangeDatum(&ran.HighVal[i], lengths[i])
+			fixRangeDatum(&ran.HighVal[i], lengths[i], tp[i])
 		}
 		ran.HighExclude = false
 	}
 }
 
-func fixRangeDatum(v *types.Datum, length int) {
+func fixRangeDatum(v *types.Datum, length int, tp *types.FieldType) {
 	// If this column is prefix and the prefix length is smaller than the range, cut it.
-	if length != types.UnspecifiedLength && length < len(v.GetBytes()) {
-		v.SetBytes(v.GetBytes()[:length])
+	// In case of UTF8, prefix should be cut by characters rather than bytes
+	if v.Kind() == types.KindString || v.Kind() == types.KindBytes {
+		colCharset := tp.Charset
+		if colCharset == charset.CharsetUTF8 || colCharset == charset.CharsetUTF8MB4 {
```

Comment:
how about:
```go
colValue := v.GetBytes()
isUTF8Charset := colCharset == charset.CharsetUTF8 || colCharset == charset.CharsetUTF8MB4
if isUTF8Charset && length != types.UnspecifiedLength && utf8.RuneCount(colValue) > length {
	...
} else if length != types.UnspecifiedLength && len(colValue) > length {
	v.SetBytes(colValue[:length])
}
```

Reply:
@zz-jason Actually I just copied the same logic from https://github.com/pingcap/tidb/blob/e0034f959a54ff58a24eea744d5cc5a095da0454/table/tables/index.go#L133
I left it the same because it seems hard to decide where to put the reusable code. Should both logic be changed or I could move onto another PR to solve this?

Reply:
Uh, I think we can change both logic here in this PR.

Reply:
Done. In addition, isUTF8Charset must be on the outer if-statement to maintain the correct logic.